scripts/bumper/main.go

```diff
@@ -0,0 +1,320 @@
+			if len(commitMatches[0]) > 1 {
+				lastCommit = string(commitMatches[1])
+			}
+		}
+		if lastCommit != "" {
+			logger.WithField("commit", lastCommit).Debug("found last commit synchronized with staging")
+			lastCommits[path] = lastCommit
+		}
+
+		if path != "." {
+			return fs.SkipDir
+		}
+		return nil
+	}); err != nil {
+		return nil, fmt.Errorf("failed to walk %s: %w", stagingDir, err)
+	}
+
+	var commits []commit
+	for repo, lastCommit := range lastCommits {
+		if _, err := runCommand(logger, exec.CommandContext(ctx,
+			"git", "fetch",
+			"git@github.com:operator-framework/"+repo,
+			"master",
+		)); err != nil {
+			return nil, err
+		}
+
+		output, err := runCommand(logger, exec.CommandContext(ctx,
+			"git", "log",
+			"--pretty=%H",
+			lastCommit+"...FETCH_HEAD",
+		))
+		if err != nil {
+			return nil, err
+		}
+
+		for _, line := range strings.Split(output, "\n") {
+			line = strings.TrimSpace(line)
+			if line != "" {
+				infoCmd := exec.CommandContext(ctx,
+					"git", "show",
+					line,
+					"--pretty=format:%H|%cI|%an|%s",
+					"--quiet",
+				)
+				stdout, stderr := bytes.Buffer{}, bytes.Buffer{}
+				infoCmd.Stdout = &stdout
+				infoCmd.Stderr = &stderr
+				logger.WithField("command", infoCmd.String()).Debug("running command")
+				if err := infoCmd.Run(); err != nil {
+					return nil, fmt.Errorf("failed to run command: %s %s: %w", stdout.String(), stderr.String(), err)
+				}
+				parts := strings.Split(stdout.String(), "|")
+				if len(parts) != 4 {
+					return nil, fmt.Errorf("incorrect parts from git output: %v", stdout.String())
+				}
+				committedTime, err := time.Parse(time.RFC3339, parts[1])
+				if err != nil {
+					return nil, fmt.Errorf("invalid time %s: %w", parts[1], err)
+				}
+				commits = append(commits, commit{
+					Hash:    parts[0],
+					Date:    committedTime,
+					Author:  parts[2],
+					Message: parts[3],
+					Repo:    repo,
+				})
+			}
+		}
+	}
+	sort.Slice(commits, func(i, j int) bool {
+		return commits[i].Date.Before(commits[j].Date)
+	})
+	return commits, nil
+}
+
+func cherryPick(ctx context.Context, logger *logrus.Entry, c commit) error {
+	{
+		output, err := runCommand(logger, exec.CommandContext(ctx,
+			"git", "cherry-pick",
+			"--allow-empty", "--keep-redundant-commits",
+			"-Xsubtree=staging/"+c.Repo, c.Hash,
+		))
+		if err != nil {
+			if strings.Contains(output, "vendor/modules.txt deleted in HEAD and modified in") {
+				// we remove vendor directories for everything under staging/, but some of the upstream repos have them
+				if _, err := runCommand(logger, exec.CommandContext(ctx,
+					"git", "rm", "--cached", "-r", "--ignore-unmatch", "staging/"+c.Repo+"/vendor",
+				)); err != nil {
+					return err
+				}
+				if _, err := runCommand(logger, exec.CommandContext(ctx,
+					"git", "cherry-pick", "--continue",
+				)); err != nil {
+					return err
+				}
+			} else {
+				return err
+			}
+		}
+	}
+
+	for _, cmd := range []*exec.Cmd{
```

Comment:
A bit concerned here about the difference between this and `scripts/sync_pop_candidates.sh` (plus `sync.sh`). Some of it has to do with direct invocation of some of these commands, which might be better suited to do via `make`.

The sync scripts do:

* `make vendor` (which does tidy/vendor/verify)
* `make verify-manifests` (which does `make manifests` with `OLM_VERSION` set, which I think is a better way to do it)
*  `make verify-nested-vendor` (which may not be necessary)
*  `make verify-commits` (which is not done here, and probably should be "just in case", at least initially)

This lacks the recovery mechanisms in the the scripts, but those may not be needed if this is run frequently enough.

If this is to replace the sync script, then the sync scripts should be modified to build and call this instead.

I'm also not seeing a new branch created to run this in; does it do it on the local `master` branch?

Reply:
> Some of it has to do with direct invocation of some of these commands, which might be better suited to do via make.

I was getting weird segfaults doing that and it's a minimal risk I don't really think matters.

> The sync scripts do:

I don't want this to do that. This syncs. If you want to verify, you can do that yourself. An automated syncer does the syncing, and pushes that up to get CI. If CI fails, we see it.

> If this is to replace the sync script

Give me some time. This will be improved and made automatic soon enough.

> I'm also not seeing a new branch created

Again, I don't want it to. If you want a new branch, do that. This tool works in the current context. The libraries that I'll tap into for publishing this as a PR and all that expect as much. I also (personally) prefer controlling this externally.

Reply:
Consider that if we run the `make verify*` bits, we will need to write code here to disambiguate between "we failed to generate new manifests" and "the verification of manifests failed, since we're just generating them for the first time and expect to have a git diff". Not really clear to me how that is better than running `make manifests` directly.

Reply:
> sync scripts should be updated to use this (if this goes in)

My intent is to deliver a series of small PRs that build this up into the fully-fledged codebase it will become. I hope we can accommodate this ... ? I was using this proto-bumper already and figured it could stand alone while the rest got implemented.

Reply:
To be clear, the risk here is of a pretty small magnitude. If `make verify-manifests` ever changes what it expects, and we did not remember to update this script, the next automated sync PR will fail tests. I'll add a `bump-manifests` target since that will be useful to individual devs as well.

Reply:
Which is why I'm suggesting `bump-manifests`